src/Store.ts

```diff
@@ -0,0 +1,581 @@
+	queryLog: dstore.QueryLogEntry<any> [];
+
+	storage: Evented;
+
+	// stringify: Function
+	//		For stores that serialize data (to send to a server, for example) the stringify
+	//		function can be specified to control how objects are serialized to strings
+	stringify: () => any;
+
+	constructor(options?: StoreArgs) {
+		super();
+		this.autoEmitEvents = true;
+		this.autoEmitHandles = [];
+		this.idProperty = 'id';
+		this.queryLog = [];
+		this.Filter = Filter;
+		// Set defaults
+		// perform the mixin
+		options && lang.mixin(this, options);
+		// TODO - Don't do this, it's a workaround to make inheritance from this class via
+		// declare work properly.
+		for (let name in this) {
+			let value = this[name];
+			if (Object.prototype.toString.call(value) === '[object Function]') {
+				value.nom = name;
+			}
+		}
+
+		// TODO - This is designed specifically for working with declare-based inheritance and needs to be reworked.
+		if (this.Model && this.Model.createSubclass) {
+			// we need a distinct model for each store, so we can
+			// save the reference back to this store on it.
+			// we always create a new model to be safe.
+			this.Model = this.Model.createSubclass([]).extend({
+				// give a reference back to the store for saving, etc.
+				_store: this
+			});
+		}
+
+		// the object the store can use for holding any local data or events
+		this.storage = new Evented();
+		const store = this;
+		if (this.autoEmitEvents) {
+			// emit events when modification operations are called
+			this.autoEmitHandles.push(aspect.after(this, 'add', <any> this._emitUpdateEvent('add')));
+			this.autoEmitHandles.push(aspect.after(this, 'put', <any> this._emitUpdateEvent('update')));
+			this.autoEmitHandles.push(aspect.after(this, 'remove', function (result, args) {
+				const emit = function (result: Promise<any> | any) {
+					store._emit('delete', {
+						id: args[0],
+						type: undefined
+					});
+
+					return result;
+				};
+
+				if (result && result.then) {
+					result.then(emit);
+					return result;
+				}
+
+				return emit(result);
+			}));
+		}
+	}
+
+	private _emit(type: string, event: {type: string, cancelable?: boolean}): void | boolean {
+		event = event || {
+			type: undefined
+		};
+		event.type = type;
+		try {
+			return this.storage.emit(event);
+		} finally {
+			// Return the initial value of event.cancelable because a listener error makes it impossible
+			// to know whether the event was actually canceled
+			return event['cancelable'];
+		}
+	}
+
+	protected _emitUpdateEvent(type: string) {
+		return function (result: Promise<any> | any, args: {beforeId: string | number}[]) {
+			const self = this;
+			const emit = function (result: Promise<any> | any) {
+				const event: Hash<any> = {
+					target: result
+				};
+				const options = args[ 1 ] || <any> {};
+				if ('beforeId' in options) {
+					event['beforeId'] = options.beforeId;
+				}
+				self._emit(type, event);
+				return result;
```

Comment:
The return wasn't necessary. But calling emit before calling then when it is a promise, and not calling emit otherwise both cause other tests to fail, so the inner function still seems to be necessary.